### ISSUE TITLE
00:41 Info  DuplicateCheckerPlugin - Event submission cancelled by plugin: refid= type= message=MYMESSAGE

### ISSUE THREAD

**Issue by MihaMarkic** [exceptionless/Exceptionless.Net#147]
Who is this DuplicateCheckerPlugin guy and what does it want from me?

> 00:41 Info  DuplicateCheckerPlugin - Event submission cancelled by plugin: refid= type= message=MYMESSAGE

This is from UseInMemoryLogger and Exceptionless indeed doesn't send anything.

**Comment by MihaMarkic** [comment]
Also note that my app is stuck with this error. Restart doesn't help. It also doesn't help that this error is silent (normally you wouldn't notice it unless logging is activated).

**Comment by niemyjski** [comment]
@MihaMarkic We have a plugin (DuplicateCheckerPlugin) that sits in the pipeline and removes duplicate events. So for example event1 comes through we keep a hash of the entire event and mark it as sent. Lets say event2 comes in which is exactly the same as event 1, we cancel it and store it in a cache for 60 seconds. Then for the next 60 seconds any event that is exactly the same in every way is cancelled and updates event2's date and count property, so we know how many and when the last one has occurred. Then after the 60 second window we submit event2 with the exact count of how my occurred.

**Comment by niemyjski** [comment]
So if you wait enough (60 seconds) you should see exact counts in the ui for that error.

**Comment by niemyjski** [comment]
Are you running into an error where you believe it's discarding non unique events? Are you seeing the rolled up events in the ui?

**Comment by MihaMarkic** [comment]
There is definitely something odd going on. Trying to pinpoint but it is like this:
Sometimes it won't send anything out, log would be empty (except for a single duplication warning). I produce like 20-30 logs in few seconds from more than one thread.
ExceptionlessClient.Default.SubmitLog is thread safe, right?

**Comment by MihaMarkic** [comment]
It is a .net 4.6.1 console application.

**Comment by MihaMarkic** [comment]
Also note that I issue ExceptionlessClient.Default.ProcessQueue(); before the end of the application.

**Comment by niemyjski** [comment]
Yes, it should be thread safe. If you want, I would be more than happy to go on a remote meeting and take a look at this with you.

**Comment by MihaMarkic** [comment]
@niemyjski I think I know what's going on. 

TL;DR; ProcessQueue() doesn't flushes always.

Rationale (I went looking into source code).
ProcessQueue would exit immediately when DefaultEventQueue.Process is running (it sets _processingQueue=true).
So, what's happening to me is that internal Timer fires DefaultEventQueue.Process and during this same time my application ends and I call ExceptionlessClient.Default.ProcessQueue(); which exits immediately (because of processingQueue flag) even though queue is being processed and it is not empty.

Fix: Not trivial, I guess. ExceptionlessClient.Default.ProcessQueue() should block until queue is empty. ProcessQueue should check if queue is empty before it exits. And if not empty, then it should wait for Process to do its job by using some synchronization.

Makes sense.

**Comment by MihaMarkic** [comment]
Thinking, perhaps an easy solution would be to convert _processingQueue from bool to an ManualResetEvent or something.
Also _processingQueue doesn't feel like very thread safe since it isn't using any synchronization at all.

**Comment by niemyjski** [comment]
Thanks for the pr, greatly appreciate it. The nightlies are available on myget. I'm going to do some testing but also need to make some csproj changes before the next release. If you find any issues please let us know.
